onnxruntime/core/providers/coreml/builders/impl/split_op_builder.cc

```diff
@@ -43,55 +46,104 @@ Status SplitOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
-    // note: checked in IsOpSupportedImpl that ensures the dim value at splitting axis exists
-    auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
-    uint64_t chunk_size = narrow<uint64_t>((split_dim_size + num_outputs - 1) / num_outputs);
-    uint64_t remainder = split_dim_size % chunk_size;
-    if (remainder) {
-      // uneven
-      auto split_sizes = InlinedVector<uint64_t>(num_outputs, chunk_size);
-      split_sizes.back() = remainder;
-      for (size_t i = 0; i < split_sizes.size(); i++) {
-        coreml_splitnd->add_splitsizes(split_sizes[i]);
+  int64_t axis = helper.Get("axis", 0);
+
+
+#if defined(COREML_ENABLE_MLPROGRAM)
+  if (model_builder.CreateMLProgram()) {
+      int64_t num_outputs;
+      using namespace CoreML::Specification::MILSpec;
+      std::unique_ptr<Operation> split_op = model_builder.CreateOperation(node, "split");
+      AddOperationInput(*split_op, "axis", model_builder.AddScalarConstant(split_op->type(), "axis", axis));
+
+      if (input_defs.size() > 1) {
+        // if "split" is explicitly provided as an input
+        const auto& split_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());
+        Initializer unpacked_tensor(split_tensor);
+        auto split_span = unpacked_tensor.DataAsSpan<int64_t>();
+        num_outputs = narrow<int64_t>(split_span.size());
+        AddOperationInput(*split_op, "split_sizes", model_builder.AddConstant(split_op->type(), "split_sizes", split_span));
+      } else if (node.SinceVersion() < 18) {
+        num_outputs = narrow<uint64_t>(node.OutputDefs().size());
+        AddOperationInput(*split_op, "num_splits",
+            model_builder.AddScalarConstant(split_op->type(), "num_splits", num_outputs));
+      } else {
+        // note: for opset 18+ 'num_outputs' is a required attribute
+        num_outputs = narrow<int64_t>(helper.GetInt64("num_outputs").value());
+        // note: checked in IsOpSupportedImpl that ensures the dim value at splitting axis exists
+        auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
+        uint64_t chunk_size = narrow<int64_t>((split_dim_size + num_outputs - 1) / num_outputs);
+        uint64_t remainder = split_dim_size % chunk_size;
+
+        if (remainder) {
+          // uneven
+          std::vector<int64_t> split_sizes(num_outputs, chunk_size);
+          split_sizes.back() = remainder;
+          AddOperationInput(*split_op, "split_sizes", model_builder.AddConstant(split_op->type(), "split_sizes", split_sizes));
+
+        } else {
+          // even
+          AddOperationInput(*split_op, "num_splits",
+            model_builder.AddScalarConstant(split_op->type(), "num_splits", num_outputs));
+        }
+
+        AddOperationInput(*split_op, "x", input_defs[0]->Name());
+        for (int64_t i = 0; i < num_outputs; i++) {
+          AddOperationOutput(*split_op, *node.OutputDefs()[i]);
+        }
```

Comment:
nit: the actual number of entries in OutputDefs() should be == num_outputs for the onnx model to be valid so we could potentially simplify and not worry about passing num_outputs through to here (e.g. delete line 64)